[#3] Player Movement 기능 및 애니메이션 구현

JunHyeong777 opened the pull request:
close #3 

**PlayerMovementTest Scene에서 작업함**
테스트 Ground 오브젝트의 길이를 늘렸고 "Ground"태그 지정, PlayerData.json에서 speed 값 5로 변경함
Unity Asset Store의 [Dragon Warrior (Free)](https://assetstore.unity.com/packages/2d/characters/dragon-warrior-free-93896#contentl) 에셋 사용함

- 점프, 더블 점프 구현
  - 스페이스 바 누르면 점프
  - 점프력 및 점프 조작감은 RigidBody 2D의 Gravity Scale 값과 Player Controller의 jumpForce 값을 변경하여 조정 가능 
  - 슬라이딩 중에 점프 시 슬라이딩을 중지

- 슬라이드 기능 구현
  - Left Shift 키를 누르는 동안 슬라이드 상태 
  - 점프 중에는 슬라이드 불가능하도록 했음
 

- 애니메이션 추가
  - run
  ![run](https://github.com/Famec7/OpenSourceProject/assets/103454894/37bb089a-e6e7-4512-8743-7ab24239923c)
  - jump
  ![jump](https://github.com/Famec7/OpenSourceProject/assets/103454894/d61928bc-a694-4b90-8a35-ebb5a9a25678)
  - slide
  ![slide](https://github.com/Famec7/OpenSourceProject/assets/103454894/97284b24-f56d-4911-9744-d11a73e12bdc)

 
PlayerController
```
using UnityEngine;

public class PlayerController : MonoBehaviour
{
    private PlayerData _playerData = null;
    private Animator _animator;
    private Rigidbody2D _rigidbody2D;
    private BoxCollider2D _collider;

    // 점프 관련 변수
    public float jumpForce = 10.0f;  // 점프 시 적용될 힘 - 점프 힘과 RigidBody 2D의 Gravity Scale 값을 조정하여 점프 조정 가능   
    private int _jumpCount = 0;      // 현재 점프 횟수
    private int _maxJumpCount = 2;   // 최대 점프 횟수 (더블 점프)

    // 콜라이더 크기 및 오프셋
    private Vector2 _standingColliderSize;
    private Vector2 _standingColliderOffset;
    public Vector2 slidingColliderSize;     // 슬라이딩 시 콜라이더 크기
    public Vector2 slidingColliderOffset;   // 슬라이딩 시 콜라이더 오프셋

    // 플레이어 상태를 나타내는 열거형
    enum PlayerState 
    {
        Running,
        Jumping,
        Sliding
    }

    private PlayerState _currentState = PlayerState.Running; // 초기 상태는 Running

    private void Awake()
    {
        _playerData = DataManager.Instance.LoadData();
        _animator = GetComponent<Animator>();
        _rigidbody2D = GetComponent<Rigidbody2D>();
        _collider = GetComponent<BoxCollider2D>();
    }

    private void Start()
    {
        // 기존 콜라이더 크기 및 오프셋 저장
        _standingColliderSize = _collider.size; 
        _standingColliderOffset = _collider.offset; 

        // 슬라이딩 시 적용될 콜라이더 및 오프셋 설정 (기존 크기의 절반)
        slidingColliderSize = new Vector2(_collider.size.x, _collider.size.y / 2);
        slidingColliderOffset = new Vector2(_collider.offset.x, _collider.offset.y - _collider.size.y / 4);
    }

    private void Update()
    {
        // 점프 입력 처리
        if (Input.GetButtonDown("Jump") && _jumpCount < _maxJumpCount)
        {
            Jump();
        }

        // 슬라이딩 입력 처리
        if (Input.GetKey(KeyCode.LeftShift) && _currentState == PlayerState.Running)
        {
            StartSlide();
        }
        else if (Input.GetKeyUp(KeyCode.LeftShift))
        {
            StopSlide();
        }
    }

    private void Jump()
    {
        // 슬라이딩 중에 점프 시 슬라이딩을 중지
        if (_currentState == PlayerState.Sliding)
        {
            StopSlide();
        }

        // 점프 로직 수행
        _rigidbody2D.velocity = new Vector2(_rigidbody2D.velocity.x, jumpForce);
        _jumpCount++;
        _currentState = PlayerState.Jumping;
        _animator.SetBool("Jump", true); // 점프 애니메이션 활성화
    }


    // 슬라이딩 상태로 전환
    private void StartSlide()
    {
        _currentState = PlayerState.Sliding;
        _collider.size = slidingColliderSize;
        _collider.offset = slidingColliderOffset;
        _animator.SetBool("Slide", true); // 슬라이딩 애니메이션 활성화
    }

    // 슬라이딩 상태 해제
    private void StopSlide()
    {
        _animator.SetBool("Slide", false);
        _currentState = PlayerState.Running;
        _collider.size = _standingColliderSize;
        _collider.offset = _standingColliderOffset;
    }

    private void OnCollisionEnter2D(Collision2D collision)
    {
        if (collision.gameObject.CompareTag("Ground"))
        {
            _animator.SetBool("Jump", false);
            _jumpCount = 0; // 땅에 닿으면 점프 카운트 초기화
            _currentState = PlayerState.Running; // 땅에 닿으면 상태를 Running으로 전환
        }
    }
}
```

JunHyeong777 committed:
feat: 1단 점프 & 2단 점프 구현 완료

1. Update 메소드 내에서 "Jump"입력을 감지해 점프하도록 구현
2. jumpCount 변수로 현재 점프 횟수를 계산하고 maxJumpCount로 점프 한계 횟수를 지정 (현재 2회)
3. Ground 태그를 가진 오브젝트와 충돌하면 jumpCount를 0으로 초기화
4. 테스트 Ground 오브젝트의 길이를 늘렸고 "Ground"태그 지정, PlayerData.json에서 speed 값 5로 변경함

JunHyeong777 committed:
feat: 슬라이딩 구현 완료

1. Left Shift 키를 누르는 동안 슬라이딩하도록 구현
2. Circle 오브젝트 이름 Player로 변경하고 Circle Collider에서 Box Collider 2D로 변경했음
3. 슬라이딩 시 Player의 Box Collider 크기가 위아래로 절반이 되도록 구현했음 (콜라이더 오프셋도 이에 맞게 조정됨)

JunHyeong777 committed:
feat: 애니메이션 구현 완료

1. Player의 run, jump, slide 애니메이션 추가
2. unity asset store의 Dragon Warrior (Free) 에셋 사용
3. 에셋 PNG 파일들 피벗 콜라이더에 맞게 조정
 - flyKick 파일들 제외하고 모두 X 0.5, Y 0.25로 조정함
4. 점프 중에 공중에서 슬라이드 하지 못하도록 변경

Famec7 reviewed (commented):
- GetComponent는 Start보다는 Awake에서 호출하는 것이 좋을 것 같습니다
- 슬라이드랑 점프 동시 사용이 불가능하면 bool보다는 enum으로 관리하는 것은 어떤가요??
- 주석 조금만 더 추가하면 좋을 듯 싶습니다!! (처음 본 사람도 코드를 알아야하기 때문에...)

JunHyeong777 committed:
refactor: PlayerController 리팩토링

1. GetComponent Start에서 Awake에서 호출하도록 변경
2. Player의 상태를 bool에서 enum으로 관리하도록 변경
3. 주석 추가

JunHyeong777 commented:
> * GetComponent는 Start보다는 Awake에서 호출하는 것이 좋을 것 같습니다
> * 슬라이드랑 점프 동시 사용이 불가능하면 bool보다는 enum으로 관리하는 것은 어떤가요??
> * 주석 조금만 더 추가하면 좋을 듯 싶습니다!! (처음 본 사람도 코드를 알아야하기 때문에...)

모두 반영해서 "[refactor: PlayerController 리팩토링](https://github.com/Famec7/OpenSourceProject/pull/6/commits/da2e92ba89d11715b2d767c8b8c4e667e36f7c46)"  커밋 했습니다.

Famec7 commented:
코드에 이상없어서 머지하겠습니다

Famec7 reviewed (approved):
코드에 이상 없어서 머지하겠습니다
